### ISSUE TITLE
Model with other backbone networks?

### ISSUE THREAD

**Issue by rocketbear** [ZheC/Realtime_Multi-Person_Pose_Estimation#131]
Hi, everyone.

I am trying to re-implement the pose estimation model using other backbone networks, e.g. mobile-net, res-50, etc. Currently I have no success in using other backbone networks. For example, when using only state-1 and stage-2 and mobile-net pretrained on imagenet, the loss stopped at about 370. But when I use the first 10 conv layers of VGG-19 pretrained on imagenet as suggested in the paper, the training error converges to <200 quickly. But the training loss stopped at 370 when using VGG-19 but without the pretraining weights.

I wonder if anyone has successfully trained the model using other types of backbone networks. Any advices will be greatly appreciated!

**Comment by rocketbear** [comment]
The convergence problem is solved by using residual connections all through the model, i.e. in both the backbone feature extraction layers and the additional cpm layers. FYI.

**Comment by ZheC** [comment]
Thank you for the updates, this can help other people trying the same thing!

**Comment by anatolix** [comment]
@rocketbear could you please commit it somewhere to look?

**Comment by Xiangyu-CAS** [comment]
@rocketbear Hi, I have tried to use many different backbones, such as VGG-16, ResNet-50 and DenseNet-121. It works fine when using DenseNet and VGG-16 but failed with ResNet-50. The performance of ResNet-50 is terrible, about 40% mAP. I also tried to add residual connections in CPM layers, but  turned out worse. Is this possible for you to share the training prototxt file?

**Comment by Ai-is-light** [comment]
@rocketbear , Hi ,you said , 'the training error converges to <200 quickly', I would like to know, the training error you said is the total loss of the stage1 to stage 6 , and the total loss of L1  and  L2 loss? Another question is about the backbone feature extraction layers and the additional cpm layers. You said ,'The convergence problem is solved by using residual connections all through the model, i.e. in both the backbone feature extraction layers and the additional cpm layers. FYI.' 
The backbone feature extraction model you chose is the DenseNet or the mobileNet? I have tried the mobileNet for training the faster and smaller model, but it doesn't work , and the loss is as high as 1000 of total 6 stages.  Furthermore, about the additional cpm layers, during my training and testing the cpm branch can work well, the paf can not work well and always trends toward to disconvergence .
I mean, would you mind sharing the additional cpm layers design and how do you deal with the paf branch, keeping it as the original?
Thanks, looking forward  to your kindness answers

**Comment by Ai-is-light** [comment]
@rocketbear and, would you mind asking us how fast about your model?

**Comment by rocketbear** [comment]
The goal of my work is to train a much faster model with user acceptable accuracy on our own data. Since there are only a few difficult poses in our data, I decide to significantly reduce the model size. So my model has only 2 stages, with 1/5 the FLOPS of the first two stages of the author's model, and runs 4 to 5 times faster on both cpu and gpu. 

The prototxt of the training net and the solver can be found at [https://github.com/rocketbear/pose_estimation](url)
Note that I have re-implemented our customized data layer.

The affinity prediction and joint prediction shares all convolution feature maps, i.e. there is only one branch up to the prediction output. 3*3 conv is used in stage 2, but with more layers than the author's model. Residual modules are used in both stage 1 and stage 2. The res module used in my model contains two 3*3 convolutions instead of 3*3-1*1-3*3. I have also tried mobilenet but failed to get a good model. The training loss stopped at 310 for my 2-stage model, so it would be around 1000 if all 6 stages are included, which is similar to @Ai-is-light 's result.

The model is trained from scratch with batch size =40, all weights are initialized with Gaussian std=0.01, and all biases are initialized to 0. The initial learning rate is 1e-4, and is divided by 10 every 100k iterations. The training loss converges to about 185 after 300k iterations. If 4 more stages were added (like the author's 6-stage model), the training loss should be around 500, since the later stages should have lower loss than the earlier stages.

 I have not evaluate the accuracy on the COCO dataset yet, but the results on several sample images look similar to that generated by the first two stages of the author's model.

**Comment by Ai-is-light** [comment]
Thanks.I also try to get the small and faster model @rocketbear , and failed when I tried to do it by using the mobileNet. I got a model as half as author and as 2x faster than author
![wechatimg274](https://user-images.githubusercontent.com/11785119/35909817-86cbbaae-0ba9-11e8-8fb4-45417f38a21c.jpeg)
. However, I just try the 6-stages model. During the my practice, I have found that the author's model with large redundancy, I also found that the stage1's loss reduced firstly, and then stage6 for the first 10k to 30k iterations. But, more importantly and interesting , the stage3's loss are the last one to converge, and the author @ZheC showed  the loss-L1 curve and loss-L2 curve just about the **stage 3**, **# **why****? I just wonder that during data through the stage3-stage6, the network trend to overfitting, although 6 stages network is just for the training for the coarse to fine and get higher mAP. And, to be honest, the model I trained can tell me  the network starting to converge well at iterations about 100 - 300k.When  I test the model at 300K and 440K , the mAP is nearly the same and the loss of the total 6 stages(L1 and L2) is about 450-500. so, I guess the stage3-stage6 can be changed to a deeper and wider network than the one stage

**Comment by Ai-is-light** [comment]
I set the bitch size is 5 , and 4 gpus

**Comment by mkocabas** [comment]
Hi, I can succesfully tried ResNet50 as backbone and it improved the accuracy a lot compared to author. However I used ResNet50 with dilated convolutions, by this way input image downsampled 8x times compared to 32x times in original ResNet. Also you should freeze batchnorm layers as imagenet trained parameters.

**Comment by Xiangyu-CAS** [comment]
@mkocabas Hi, I  did exactly as what you said  but failed. ResNet-50 with dilation was hard to converage and ResNet-101 with dilation was much worse than vgg. 
Could you possiblely share your prototxt with me? 

Thanks a lot!

**Comment by Ai-is-light** [comment]
@mkocabas would mind sharing the setting of the stage1 to stage6, they are also the Res block with dilated convolutions, or other special changes? And, what's more, I also would like to know how fast of your model you trained , what is the size of your model, how long will your networks need to converge with the loss no decrease.
Thanks, looking forward your any suggestion

**Comment by Xiangyu-CAS** [comment]
Hi,  My network with ResNet-50 can successfully work now. To save time, I only use 8000 images from COCO_train and test it on 50 images. Evaluation result improved from 48%(vgg) to 51%(resnet).

[https://github.com/Xiangyu-CAS/Pose-Estimation/blob/master/pose_train_test.prototxt](url)

However, What I observed is totally different from @rocketbear and @mkocabas . 
1. Residual connection is not used in CPM, all the thing I did is to replace VGG-19 with ResNet-50
2. I did not freeze BN. Actually, when I freeze BN, the network became hard to converge and the performance is extremely low (8%). On the contrary, open BN make it improved from 8% to 51%

I think my conclusion with BN is weird, since all the papers have mentioned freezing BN. I would appreciate it if anyone could point out what is wrong with my configuration.

**Comment by Brave731** [comment]
I try you prototxt file, but get result mAP46%. Can you tell me training details, such as iter, optimizer.

**Comment by Brave731** [comment]
@Xiangyu-CAS I try you prototxt file, but get result mAP46%. Can you tell me training details, such as iter, optimizer.

**Comment by RayLeijingsi** [comment]
> Hi, My network with ResNet-50 can successfully work now. To save time, I only use 8000 images from COCO_train and test it on 50 images. Evaluation result improved from 48%(vgg) to 51%(resnet).
> 
> [https://github.com/Xiangyu-CAS/Pose-Estimation](url)
> 
> However, What I observed is totally different from @rocketbear and @mkocabas .
> 
> 1. Residual connection is not used in CPM, all the thing I did is to replace VGG-19 with ResNet-50
> 2. I did not freeze BN. Actually, when I freeze BN, the network became hard to converge and the performance is extremely low (8%). On the contrary, open BN make it improved from 8% to 51%
> 
> I think my conclusion with BN is weird, since all the papers have mentioned freezing BN. I would appreciate it if anyone could point out what is wrong with my configuration.

Hello, I also want to choose some images from COCO, to build a small dataset. But I have no idea about annotations-keypoint.json. Would you please share your experiences with me?
